### PR TITLE
refactor: various code improvements + add missing tests

### PR DESCRIPTION
**Pull request by craigzour**
closes https://github.com/cds-snc/platform-forms-client/issues/4278

# Summary | Résumé

- Brings in several code improvements and new abstractions
  - `cache.ts`: new generic implementation to cache data in Redis
  - `zitadelConnector.ts`: new connector to hide communication protocol enforced by Zitadel API and follow our pattern of having third party services contained in the `integration` folder
  - `awsSqsQueueLoader.ts`: extracted logic to load SQS queue url to simplify audit logs unit tests
  - Miscellaneous:
    - Converted some occurrences of `async/await` to `promise.then` pattern
    - Introduced new `logging`, `encryption` and `utils` folders
    - "Preload" some resources in the `server.ts` > `server.listen` callback function to avoid hiding things like that in specific files where they would sometimes need to be feature flag to not run in test mode.
    - Simplify error logging by leveraging Pino API
- Adds missing unit tests
  - `encryptFormSubmission.test.ts`
  - `getPublicKey.test.ts`
  - `awsSqsQueueLoader.test.ts`
  - `encryptFormSubmission.test.ts`
  - `zitadelConnector.test.ts.test.ts`
  - `auditLogs.test.ts`
  - `cache.test.ts`

### CODE REVIEW

File: src/lib/integration/redisConnector.ts [hunk 1]
```diff
@@ -41,16 +42,16 @@ export class RedisConnector {
...
         RedisConnector.instance.client.connect();
-      await RedisConnector.connectionPromise;
-    } else {
-      // Ensure all calls to getInstance() wait for the connection to be initialized before returnign the Redis Instance
-      await RedisConnector.connectionPromise;
     }
-    return RedisConnector.instance;
+
+    // Ensure all calls to getInstance() wait for the connection to be initialized before returning the Redis Instance
+    return RedisConnector.connectionPromise.then((_) => {
```
Comment (bryan-robitaille) [inline src/lib/integration/redisConnector.ts:53]: There's no need to declare the variable if it'll be unused.
```suggestion
    return RedisConnector.connectionPromise.then(() => {
```

File: src/lib/logging/auditLogs.ts [hunk 2]
```diff
@@ -22,26 +23,7 @@ export enum AuditSubjectType {
...
-  }
-  return queueUrlRef;
-};
-
-//Initialise the queueUrlRef on load except when running tests
-// This mock should be refactored when we once we have a better way to mock the AWS SDK
-process.env.NODE_ENV !== "test" && getQueueUrl();
-
-export const logEvent = async (
+export const publishAuditLog = (
```
Comment (bryan-robitaille) [inline src/lib/logging/auditLogs.ts:26]: Any specific reason for renaming the function?  My thought was to keep the function name similar between the App and Api code bases.

**Review by bryan-robitaille — COMMENTED** [review]
Let's meet and discuss because I've noticed a lot of changes throughout the code base to a .then() style, removal of async declarations on functions and forced type returns. While a very clean style it might pose a barrier to onboarding other developers who are not as comfortable with async code, and chained promises.

File: src/lib/logging/auditLogs.ts (hunk 2 above)
Comment (craigzour) [reply src/lib/logging/auditLogs.ts:26]: Will rename to `auditLog`

File: src/lib/utils/cache.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  return (
+    RedisConnector.getInstance()
+      .then((redisConnector) =>
+        redisConnector.client.set(key, value, {
+          ...(expiryDelayInSeconds && {
+            EX: expiryDelayInSeconds,
+          }),
+        }),
+      )
+      // Redis client `set` operation returns something that we don't need so we convert it to `void` by calling `Promise.resolve()`
```
Comment (bryan-robitaille) [inline src/lib/utils/cache.ts:25]: Thank you for the comment explaining this.

File: src/middleware/authentication.ts [hunk 4]
```diff
@@ -1,61 +1,63 @@
...
+    next();
+  } catch (error) {
+    logMessage.error(
+      error,
+      "[middleware] Internal error while authenticating user",
     );
 
-    response.sendStatus(403);
-    return;
+    response.sendStatus(500);
```
Comment (bryan-robitaille) [inline src/middleware/authentication.ts:61]: What do you think about passing the error to the `next()` function to trigger the globalErrorHandler?  This way we can standardize all of the error handling from a single spot.

**Review by bryan-robitaille — COMMENTED** [review]
Nice work!

**Review by bryan-robitaille — APPROVED** [review]
Nice work 
I left a comment about error handling but I'll leave it up to you to decide on whether you want to implement it.

**Pull request merged**